### ISSUE TITLE
NuGet still fails to pack PCL libraries

### ISSUE THREAD

**Issue by sliekens** [NuGet/Home#1514]
Relates to: https://nuget.codeplex.com/workitem/4013
Relates to: #1106

The above issues were closed without a fix, so I'm opening a new one. Third time's the charm, right?
# tldr; nuget 2.8.8 and 3.2.0 use a broken version of AssemblyMetadataExtractor
##### The Scenario

PCL profile 151 (.NET 4.5.1, Windows 8.1, Windows Phone 8.1)

```
└───src
    │   MyPortableLib.sln
    │
    └───MyPortableLib
            MyPortableLib.csproj
            MyPortableLib.nuspec
```
##### The Command

```
nuget.exe pack src\MyPortableLib\MyPortableLib.csproj  /verbosity detailed
```
##### The Error

``` C#
WARNING: Unable to extract metadata from 'MyPortableLib.dll'.
WARNING: System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
File name: 'System.Runtime, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' ---> System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
File name: 'System.Runtime, Version=4.0.10.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at System.Reflection.RuntimeAssembly._nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.nLoad(AssemblyName fileName, String codeBase, Evidence assemblySecurity, RuntimeAssembly locationHint, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoadAssemblyName(AssemblyName assemblyRef, Evidence assemblySecurity, RuntimeAssembly reqAssembly, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean throwOnFileNotFound, Boolean forIntrospection, Boolean suppressSecurityChecks)
   at System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, Evidence assemblySecurity, StackCrawlMark& stackMark, IntPtr pPrivHostBinder, Boolean forIntrospection)
   at System.Reflection.RuntimeAssembly.InternalLoad(String assemblyString, Evidence assemblySecurity, StackCrawlMark& stackMark, Boolean forIntrospection)
   at System.Reflection.Assembly.ReflectionOnlyLoad(String assemblyString)
   at NuGet.AssemblyMetadataExtractor.MetadataExtractor.AssemblyResolver.ReflectionOnlyAssemblyResolve(Object sender, ResolveEventArgs args)
   at System.AppDomain.OnReflectionOnlyAssemblyResolveEvent(RuntimeAssembly assembly, String assemblyFullName)

WRN: Assembly binding logging is turned OFF.
To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.
Note: There is some performance penalty associated with assembly bind failure logging.
To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].

   at System.ModuleHandle.ResolveMethod(RuntimeModule module, Int32 methodToken, IntPtr* typeInstArgs, Int32 typeInstCount, IntPtr* methodInstArgs, Int32 methodInstCount)
   at System.ModuleHandle.ResolveMethodHandleInternalCore(RuntimeModule module, Int32 methodToken, IntPtr[] typeInstantiationContext, Int32 typeInstCount, IntPtr[] methodInstantiationContext, Int32 methodInstCount)
   at System.ModuleHandle.ResolveMethodHandleInternal(RuntimeModule module, Int32 methodToken, RuntimeTypeHandle[] typeInstantiationContext, RuntimeTypeHandle[] methodInstantiationContext)
   at System.Reflection.CustomAttributeData..ctor(RuntimeModule scope, CustomAttributeRecord caRecord)
   at System.Reflection.CustomAttributeData.GetCustomAttributes(RuntimeModule module, Int32 tkTarget)
   at System.Reflection.CustomAttributeData.GetCustomAttributesInternal(RuntimeAssembly target)
   at System.Reflection.RuntimeAssembly.GetCustomAttributesData()
   at System.Reflection.CustomAttributeData.GetCustomAttributes(Assembly target)
   at NuGet.AssemblyMetadataExtractor.MetadataExtractor.GetMetadata(String path)
   at NuGet.AssemblyMetadataExtractor.MetadataExtractor.GetMetadata(String path)
   at NuGet.AssemblyMetadataExtractor.GetMetadata(String assemblyPath)
   at NuGet.AssemblyMetadataExtractor.ExtractMetadata(PackageBuilder builder, String assemblyPath)
   at NuGet.Commands.ProjectFactory.CreateBuilder(String basePath)
```
##### The Solution

There is a pull request that fixes this bug that has been sitting in limbo since <strong style="font-size:200%;">2014</strong>.

https://nuget.codeplex.com/SourceControl/network/forks/StephenCleary/Issue4013/changeset/2803f4895ea678d0b0cfbb5a5e2961ec0dec40f7

This PR has been merged into the master branch, but it was never built and released.  That means that the latest NuGet.Core 2.8.8 package still has the bug, and that **both nuget 2.8.8 and 3.2.0 use a broken version of AssemblyMetadataExtractor**.

**Comment by sliekens** [comment]
Please treat with the highest priority.

Repro: https://github.com/StevenLiekens/MyPortableLib

I spent a lot of time digging around the sources to figure out why this is happening and I will personally come to your home and slap your face if you close this issue without a fix.

**Comment by yishaigalatzer** [comment]
The CodePlex repo has been abandoned for a long time now, and we do not monitor it at all (as it says on every page there). So thanks a lot! for bringing it to our attention.

On another issue I believe asked about fixing the code yourself. The NuGet core code relevant code is at https://github.com/NuGet/NuGet2/blob/dev/src/Core/Authoring/AssemblyMetadataExtractor.cs#L69

**Comment by sliekens** [comment]
I believe the code is already fixed, but not released. The command utility uses an older version of the code. This is not something I can fix.

**Comment by yishaigalatzer** [comment]
Thanks for that, we can take a look at fixing it then.

**Comment by sliekens** [comment]
Thanks. I appreciate it.

**Comment by yishaigalatzer** [comment]
Looks like @emgarten merged the fix in codeplex, but didn't port it into https://github.com/nuget/nuget2, can you please integrate?

**Comment by sliekens** [comment]
This is the file on the master branch: https://github.com/NuGet/NuGet2/blob/master/src/Core/Authoring/AssemblyMetadataExtractor.cs

It's fixed on this branch, but maybe 'master' is not the main branch anymore? The bug still exists in all of the 2.x.x branches.

**Comment by zhili1208** [comment]
merged to 2.10

**Comment by sliekens** [comment]
Thanks everyone!

**Comment by fubar-coder** [comment]
This doesn't seem to be fixed in 3.2.1rc1, is this correct?

**Comment by yishaigalatzer** [comment]
This will be in 3.3
